### PR TITLE
Rebasing Ursula's newest work.

### PR DESCRIPTION
**Pull request by emtwo**

### CODE REVIEW

File: js/grid.js [hunk 1]
```diff
@@ -67,7 +67,8 @@
       addEventListener("resize", this);
 
       gNewTab.sendToBrowser("NewTab:InitializeGrid");
-      gNewTab.registerListener("NewTab:UpdateLinks", this.refresh.bind(this));
+      gNewTab.registerListener("NewTab:InitializeLinks",
+                               this.refresh.bind(this));
```
Comment (marcoscaceres) [inline js/grid.js:71]: Nit, whitespace. jscs formatter will fix this, but as a general rule, just add two spaces.

File: js/grid.js [hunk 2]
```diff
@@ -82,10 +83,10 @@
      * @param {Cell} aCell The cell that will contain the new site.
      * @return {Site} The newly created site.
      */
-    createSite(aLink, aCell) {
+    createSite(aLink, aCell, aEnhanced) {
```
Comment (marcoscaceres) [inline js/grid.js:86]: Please avoid boolean traps. Please change to:

``` JS
const REGULAR = "regular";
const ENHANCED = "enhanced";
... 
//type defaults to "regular"
createSite(aLink, aCell, type=REGULAR) {
}

```

File: js/grid.js [hunk 3]
```diff
@@ -140,7 +142,12 @@
       let numLinks = Math.min(links.length, cells.length);
       for (let i = 0; i < numLinks; i++) {
         if (links[i]) {
-          this.createSite(links[i], cells[i]);
+          // If the link is enhanced, and enhanced is turned on, then use the enhanced link.
+          if (enhancedLinks[i] && gNewTab.enhanced) {
+            this.createSite(enhancedLinks[i], cells[i], true);
```
Comment (marcoscaceres) [inline js/grid.js:147]: See above about boolean trap.

File: js/page.js [hunk 4]
```diff
@@ -31,7 +31,8 @@
       gNewTab.registerListener("NewTab:UpdatePages", this.update.bind(this));
       gNewTab.registerListener("NewTab:PinState", this.setPinState.bind(this));
       gNewTab.registerListener("NewTab:BlockState",
-        this.setBlockState.bind(this));
+                               this.setBlockState.bind(this));
```
Comment (marcoscaceres) [inline js/page.js:34]: Nit: please use 2 spaces.

File: js/page.js [hunk 5]
```diff
@@ -105,6 +106,13 @@
       }, SCHEDULE_UPDATE_TIMEOUT_MS);
     },
 
+    sendUpdateToTest: function() {
+      setTimeout(() => {
+        let event = new CustomEvent("AboutNewTabUpdated", {bubbles: true});
```
Comment (marcoscaceres) [inline js/page.js:111]: Why does it bubble?

File: js/page.js [hunk 6]
```diff
@@ -290,17 +298,40 @@
...
+
+    /**
+     * Find the correct URL to display.
+     *
+     * @param {Object} aURL The current URL sent down from the parent.
+     */
+    findURL(aURL) {
+      for (let site of gGrid.sites) {
+        if (site && aURL.url === site.url) {
+          site._getURI(aURL);
```
Comment (marcoscaceres) [inline js/page.js:331]: Is it ok to be accessing a private member here? 

Maybe:

``` JS
gGrid.sites
  .filter(site => site && aURL.url === site.url)
  .forEach(site => site._getURI(aURL))
```

File: js/sites.js [hunk 7]
```diff
@@ -12,11 +12,12 @@
    * This class represents a site that is contained in a cell and can be pinned,
    * moved around or deleted.
    */
-  function Site(aNode, aLink) {
+  function Site(aNode, aLink, aEnhanced) {
```
Comment (marcoscaceres) [inline js/sites.js:15]: aEnhanced is another boolean trap. Can reuse suggestion above?

File: js/sites.js [hunk 8]
```diff
@@ -197,12 +198,12 @@
...
       this._checkLinkEndTime();
       // setup display variables
-      //let enhanced = Services.prefs.getBoolPref("browser.newtabpage.enhanced")
-      //                && DirectoryLinksProvider.getEnhancedLink(this.link);
-      let enhanced = gNewTab.enhanced;
+      let enhanced = this.link;
       let url = this.url;
-      let title = enhanced && enhanced.title ? enhanced.title :
-        this.link.type === "history" ? this.link.baseDomain :
+      let title = enhanced && enhanced.title &&
```
Comment (marcoscaceres) [inline js/sites.js:203]: Sorry, but is it possible to rewrite this (even though it's not your code)? nested ternary statements are super confusing :(

File: js/sites.js [hunk 9]
```diff
@@ -265,20 +266,26 @@
...
+    /**
+     * Render the correct thumbnail for the site.
+     *
+     * @param {Object} aData Contains enhanced links and the URI generated from the
+     *        current URL.
+     */
+    _getURI(aData) {
       // Only enhance tiles if that feature is turned on
-      let link = /*message.enhanced && DirectoryLinksProvider.getEnhancedLink(this.link) || */ this.link;
+      let link = aData.enhanced && this.link;
```
Comment (marcoscaceres) [inline js/sites.js:281]: Will this result in a boolean? If so, maybe `isLink`.

File: js/sites.js [hunk 10]
```diff
@@ -265,20 +266,26 @@
...
 
       let thumbnail = this._querySelector(".newtab-thumbnail");
       if (link.bgColor) {
```
Comment (marcoscaceres) [inline js/sites.js:284]: hmm.. confused here... the above makes it look like link is a boolean, but then we are checking for an attribute "bgColor"

File: js/sites.js [hunk 11]
```diff
@@ -265,20 +266,26 @@
...
         thumbnail.style.backgroundColor = link.bgColor;
       }
 
-      let uri = this.link.imageURI || ("file://" + message.uri);
+      let uri = this.link.imageURI || ("file://" + aData.uri);
```
Comment (marcoscaceres) [inline js/sites.js:288]: File URIs won't work in the context of the remote page.

File: js/updater.js [hunk 12]
```diff
@@ -191,14 +191,16 @@
...
-      Promise.all(sites.map((aSite, aIndex) => {
+      Promise.all(sites.map((aSite, aIndex, aEnhancedLinks) => {
         if (aSite || !aLinks[aIndex]) {
           return null;
         }
 
         return new Promise(resolve => {
           // Create the new site and fade it in.
-          let site = gGrid.createSite(aLinks[aIndex], cells[aIndex]);
+          let isEnhanced = aEnhancedLinks[aIndex] ? true : false;
```
Comment (marcoscaceres) [inline js/updater.js:201]: Maybe: 

``` JS
let isEnhanced = !!(aEnhancedLinks[aIndex]);
```

Comment (marcoscaceres) [inline js/updater.js:201]: Actually, if we are avoiding boolean traps: 

``` JS
let isEnhanced = aEnhancedLinks[aIndex] ? "enhanced" : "regular"; 
```

File: js/page.js [hunk 13]
```diff
@@ -290,17 +298,40 @@
...
+        gUpdater.updateGrid(message);
+      }
     },
+
+    /**
+     * Find the correct URL to display.
+     *
+     * @param {Object} aURL The current URL sent down from the parent.
+     */
+    findURL(aURL) {
```
Comment (marcoscaceres) [inline js/page.js:328]: The name of this method is a bit confusing. I would expect a `find` method to return me something. It is loading? the URL?  What does "find" mean in this context?

Comment (marcoscaceres) [inline js/page.js:328]: The name of this method is a bit confusing. I would expect a `find` method to return me something. It is loading? the URL?  What does "find" mean in this context?

File: js/page.js [hunk 14]
```diff
@@ -290,17 +298,40 @@
...
+      if (!message.pinState) {
+        gUpdater.updateGrid(message);
+      }
     },
 
     setBlockState(message) {
       for (let site of gGrid.sites) {
         if (site && site._link.url === message.link.url) {
           site._link.blockState = message.blockState;
+          break;
```
Comment (marcoscaceres) [inline js/page.js:314]: Nit: the above is `.filter()`ing... should fix that at some point to make the code a bit easier to read.

**Comment by marcoscaceres** [comment]
I gave a lot of feedback here. I'm happy to help with the changes I suggested.

File: js/sites.js (hunk 9 above)
Comment (sarracini) [inline js/sites.js:281]: aData.enhanced is a boolean, and if that's true then we want to set link to be the current link. So it's actually not a boolean :)

File: js/sites.js (hunk 10 above)
Comment (sarracini) [inline js/sites.js:284]: ^ See above comment

File: js/sites.js (hunk 9 above)
Comment (marcoscaceres) [inline js/sites.js:281]: Ok, I see. That's confusing tho :( Can we put this into a ternary or an if statement?

File: js/sites.js (hunk 10 above)
Comment (marcoscaceres) [inline js/sites.js:284]: Same as above. The above is not very idiomatic, IMO... at least, I've experienced code like this before.

**Comment by marcoscaceres** [comment]
Duplicate of #45

**Pull request closed without merging**